Jint/Native/Promise/PromisePrototype.cs

```diff
@@ -0,0 +1,204 @@
+                                    {
+                                        if (ct.Status == TaskStatus.RanToCompletion)
+                                            HandleNestedPromiseResult(ct.Result);
+
+                                        else if (ct.Status == TaskStatus.Faulted || ct.Status == TaskStatus.Canceled)
+                                        {
+                                            var rejectValue = Undefined;
+
+                                            if (ct.Exception?.InnerExceptions.FirstOrDefault() is PromiseRejectedException promiseRejection)
+                                                rejectValue = promiseRejection.RejectedValue;
+
+                                            _engine.QueuePromiseContinuation(() => chainedPromise.Reject(null, new[] {rejectValue}));
+                                        }
+
+                                    });
+                                }
+                                else
+                                    _engine.QueuePromiseContinuation(() => chainedPromise.Resolve(null, new[] {nestedResult}));
+                            }
+
+                            HandleNestedPromiseResult(result);
+                        };
+                    }
+
+                }
+                
+
+                else if (t.IsFaulted || t.IsCanceled)
+                {
+                    var rejectValue = Undefined;
+
+                    if (t.Exception?.InnerExceptions.FirstOrDefault() is PromiseRejectedException promiseRejection)
+                        rejectValue = promiseRejection.RejectedValue;
+
+                    if (rejectedCallback == Undefined)
+                    {
+                        continuation = () =>
+                        {
+                            //  If no error callback then simply pass the error value to the next promise in chain
+                            chainedPromise.Reject(null, new[] { rejectValue });
+                        };
+                    }
+                    else
+                    {
+                        continuation = () =>
+                        {
+                            rejectedCallback.Invoke(rejectValue);
+
+                            //  Chain is restored after handling catch
+                            chainedPromise.Resolve(Undefined, new[] {Undefined});
+                        };
+                    }
+                }
+
+                _engine.QueuePromiseContinuation(continuation);
+            });
+
+            return chainedPromise;
+        }
+
+        public JsValue Catch(JsValue thisValue, JsValue[] args) => Then(thisValue, new[] { Undefined, args.Length >= 1 ? args[0] : Undefined });
+
+        public JsValue Finally(JsValue thisValue, JsValue[] args)
+        {
+            if (!(thisValue is PromiseInstance promise))
+                throw ExceptionHelper.ThrowTypeError(_engine, "Method Promise.prototype.then called on incompatible receiver");
```

Comment:
should not call as throw, the intent is to have just method calls to prevent inlining problems, here doesn't matter but for consistency

Reply:
Fixed.  Unfortunately cannot use pattern matching without the redundant 'throw'.